From 94097f58f1756e4d497d2b8f66152cc73a4098f2 Mon Sep 17 00:00:00 2001
From: Yannick Fertre <yannick.fertre@foss.st.com>
Date: Thu, 31 Jul 2025 11:15:18 +0200
Subject: [PATCH] drm/stm: ltdc: re-map lcd clock tree

Share the same device tree between mp25, mp23 and mp21 family socs
(LCD clock is still tied to CK_KER_LTDC).

Change-Id: I569e6bfc35e8c6536f385ed0fffd98f21efe91e9
Signed-off-by: Yannick Fertre <yannick.fertre@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/460520
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 drivers/gpu/drm/stm/ltdc.c | 48 +++++++++++++++-----------------------
 1 file changed, 19 insertions(+), 29 deletions(-)

diff --git a/drivers/gpu/drm/stm/ltdc.c b/drivers/gpu/drm/stm/ltdc.c
index 92ccf45d6d39..57d43adb8e87 100644
--- a/drivers/gpu/drm/stm/ltdc.c
+++ b/drivers/gpu/drm/stm/ltdc.c
@@ -2652,18 +2652,11 @@ int ltdc_parse_device_tree(struct device *dev)
 int ltdc_get_clk(struct device *dev, struct ltdc_device *ldev)
 {
 	struct device_node *node;
-	int ret;
 
 	DRM_DEBUG_DRIVER("\n");
 
-	ldev->pixel_clk = devm_clk_get(dev, "lcd");
-	if (IS_ERR(ldev->pixel_clk)) {
-		if (PTR_ERR(ldev->pixel_clk) != -EPROBE_DEFER)
-			DRM_ERROR("Unable to get lcd clock\n");
-		return PTR_ERR(ldev->pixel_clk);
-	}
-
-	if (of_device_is_compatible(dev->of_node, "st,stm32mp21-ltdc")) {
+	if (of_device_is_compatible(dev->of_node, "st,stm32mp21-ltdc") ||
+	    of_device_is_compatible(dev->of_node, "st,stm32mp25-ltdc")) {
 		ldev->bus_clk = devm_clk_get(dev, "bus");
 		if (IS_ERR(ldev->bus_clk))
 			return dev_err_probe(dev, PTR_ERR(ldev->bus_clk),
@@ -2671,16 +2664,6 @@ int ltdc_get_clk(struct device *dev, struct ltdc_device *ldev)
 	}
 
 	if (of_device_is_compatible(dev->of_node, "st,stm32mp25-ltdc")) {
-		ldev->bus_clk = devm_clk_get(dev, "bus");
-		if (IS_ERR(ldev->bus_clk))
-			return dev_err_probe(dev, PTR_ERR(ldev->bus_clk),
-					     "Unable to get bus clock\n");
-
-		ldev->ltdc_clk = devm_clk_get(dev, "ref");
-		if (IS_ERR(ldev->ltdc_clk))
-			return dev_err_probe(dev, PTR_ERR(ldev->ltdc_clk),
-					     "Unable to get ltdc clock\n");
-
 		/*
 		 * The lvds output clock is not available if the lvds is not probed.
 		 * This is a usual case, it is necessary to check the node to avoid
@@ -2698,16 +2681,23 @@ int ltdc_get_clk(struct device *dev, struct ltdc_device *ldev)
 			of_node_put(node);
 		}
 
-		/*
-		 * Parent of the pixel clock should default to the reference clock (rcc clock).
-		 * If the driver has already been started, this action is not necessary and
-		 *  may cause an issue on register reading/writing.
-		 */
-		if (!device_property_read_bool(dev, "default-on")) {
-			ret = clk_set_parent(ldev->pixel_clk, ldev->ltdc_clk);
-			if (ret)
-				return dev_err_probe(dev, PTR_ERR(ldev->lvds_clk),
-						     "Could not set parent clock\n");
+		ldev->pixel_clk = devm_clk_get(dev, "ref");
+		if (IS_ERR(ldev->pixel_clk)) {
+			if (PTR_ERR(ldev->pixel_clk) != -EPROBE_DEFER)
+				DRM_ERROR("Unable to get lcd clock\n");
+			return PTR_ERR(ldev->pixel_clk);
+		}
+
+		ldev->ltdc_clk = devm_clk_get(dev, "lcd");
+		if (IS_ERR(ldev->ltdc_clk))
+			return dev_err_probe(dev, PTR_ERR(ldev->ltdc_clk),
+					     "Unable to get ltdc clock\n");
+	} else {
+		ldev->pixel_clk = devm_clk_get(dev, "lcd");
+		if (IS_ERR(ldev->pixel_clk)) {
+			if (PTR_ERR(ldev->pixel_clk) != -EPROBE_DEFER)
+				DRM_ERROR("Unable to get lcd clock\n");
+			return PTR_ERR(ldev->pixel_clk);
 		}
 	}
 
-- 
2.53.0

